Add get/set_encoding_error_handler() to control UTF-8 conversion

jmarshall opened the pull request:
Python has impressively flexible facilities to control how the [decode() function](https://docs.python.org/3/library/stdtypes.html#bytes.decode) deals with unexpected input.

The easiest way to enable scripts to control pysam's use of the `decode()` function is to expose this as a global variable via getter/setter methods, similarly to the existing `get_verbosity()`/`set_verbosity()`.

For convenience, also define an error handler that simply interprets invalid/unexpected UTF-8 bytes as ISO-8859-1 (Latin-1) characters. Make it available via `set_encoding_error_handler('pysam.latin1replace')`.

Issue #998 is about pysam's basic UTF-8 parsing refusing to parse fields containing Latin-1-encoded accented letters. This PR would allow scripts to parse such fields by adding:

```python
pysam.set_encoding_error_handler('pysam.latin1replace')
```

before calling functions that decode such fields. Other built-in standard error handlers such as `'replace'` and `'backslashreplace'` may also be of interest.

Parsing as UTF-8 and interpreting invalid input as Latin-1 is problematic because two or more adjacent Latin-1 characters can be misinterpreted as a validly UTF-8-encoded unrelated character. (Probably no such Latin-1 strings are intelligible in French or any other language.) Hence it may be better to also expose the encoding used via global getter/setters so that scripts could instead write

```python
pysam.set_encoding('latin-1')
```

to specify that their VCF v4.2 files should NOT be interpreted as UTF-8 but purely as Latin-1.

I have not added such functions because `force_str()` et al's optional arguments are currently buggy — they supply the original value of `TEXT_ENCODING` (UTF-8), not the current value, so there's little point in adding functions to change the value of `TEXT_ENCODING`. But I would be happy to rework the optional `encoding=` arguments and add such a getter/setter (probably IMHO in addition to the error handler one) if other maintainers prefer.

jmarshall commented:
Doing this via a global and getter/setter is a bit limiting.

For example, some might want to open two files simultaneously: a VCFv4.3 file to be read as UTF-8, and a CLC Workbench-written VCFv4.2 file to be read as Latin-1. So there would be some advantage in making this a per-file setting, and propagating the setting down through all the subroutines so it can be specified in each `force_str`/etc call… Not Fun. (But it might not be too bad if most of the functions that call `force_str`/etc are member methods of `VariantFile` and other file classes.)

Implementing it via a single pysam global is a lot simpler. Hopefully it provides 90% of the needed functionality…

jmarshall commented:
Renamed the functions to `get/set_encoding_error_handler()` as that's clearer and follows the precedent of some existing `get/set_encoding()` functions.

kevinjacobs-progenity commented on pysam/libcutils.pyx:
```diff
@@ -144,7 +166,7 @@ cdef charptr_to_str_w_len(const char* s, size_t n, encoding=TEXT_ENCODING):
+        return s[:n].decode(encoding, ERROR_HANDLER)
 
 
 cdef bytes charptr_to_bytes(const char* s, encoding=TEXT_ENCODING):
```
@jmarshall : Thanks again for tackling this.

Replacing the default with `None` and adding:
```
if encoding is None:
    encoding = TEXT_ENCODING
```
will allow the default to track the global variable.  This isn't so important for encoding, but the same technique will be needed to support per-file error handlers.  After this PR lands, I'd like to try to add the the error handler as an attribute to the base HTSFile class and see how hard it is to get true per-file settings.  Even if users only use the global setting API, this will also allow the error handler to be fixed at the time a file is opened, rather than changing unexpectedly as a side-effect throughout its lifetime.

jmarshall replied:
Yes indeed; an earlier version of this added an `errors=None` as an optional argument to `force_str`/etc as well. I guess the futureproofing thing to do would be to change the signatures to e.g.

```python
def charptr_to_str_w_len(const char* s, size_t n, encoding=None, errors=None):
```
and do that fixing up inside these three or four _libcutils.pyx_ functions as you suggest.

Good point about the constructor needing to give the ability to fix these fields so that file is unaffected by later changes to the global setting.

jmarshall replied:
I've updated this so all the relevant functions have `encoding=None, errors=None` optional arguments.

jmarshall commented:
My thinking on the eventual flexible API for all this is that file classes like `VariantFile` (or the base `HTSFile`, good point @kevinjacobs-progenity) should have optional constructor arguments and/or getter/setters:

```python
encoding=None
encoding_error_handler=None
```

that, if set, would override the Pysam default set by

```python
pysam.set_encoding(…)
pysam.set_encoding_error_handler(…)
```

and if you didn't use those either, the default Pysam defaults would be `utf-8` and `strict`.

jmarshall commented:
Do we like the `encoding_error_handler` terminology btw? I guess `encoding_errors`/ `get_encoding_errors`/ `set_encoding_errors` would be closer to the underlying Python `encodings` + `errors` arguments to `decode()`, but outside of that context I think making it clear that this is a handler (not e.g. a list of the actual errors encountered or something) is useful.

kevinjacobs-progenity commented:
I am happy with `encoding_error_handler`, since pysam's API is complex and has trended toward using longer and more explicit attribute and argument names.

kevinjacobs-progenity commented on pysam/libcutils.pyx:
```diff
@@ -82,7 +82,29 @@ cpdef qualities_to_qualitystring(qualities, int offset=33):
+cpdef set_encoding_error_handler(name):
+    global ERROR_HANDLER
+    previous = ERROR_HANDLER
+    if name.startswith('pysam.'):
```
Is this check intended to be more general?  It isn't obvious why you're only checking the prefix.  (Feel free to tell me to hold my horses until you're ready for it to be reviewed.)

jmarshall replied:
The user will be calling `set_encoding_error_handler()` with either a standard handler, one they wrote themselves, or one that pysam conveniently supplies. We only need to register the pysam-supplied ones if the user is actually asking for a pysam-supplied one, and in the interests of “libraries shouldn't do weird stuff if they don't need to” this check avoids registering ours if the user is using a standard one or their own.

The check could be `name == 'pysam.latin1replace'` but I'm imagining the list might expand in future. In particular, I might add a `pysam.latin15replace` as well…

That was the idea anyway… I suppose once we add file class `set_encoding_error_handler()` methods as well, it would make more sense to move this out of the global `set_encoding_error_handler` function and just always register them in whatever the appropriate form of `__init__` processing is.

kevinjacobs-progenity replied:
That makes perfect sense now that I understand where the concept is/was/may still be going.

jmarshall replied:
I've changed it so that it always just registers the handler name when _libcutils.pyx_ is loaded. (I'm no expert, but it seems like just putting the code at the top-level of the _.pyx_ file so it is executed on load is “the right form of `__init__` processing” in this context…?)

This paves the way for future file class `set_encoding_error_handler()` methods (e.g. for HTSFile) — if user code was using those, then the global `set_encoding_error_handler()` function would likely never be called (so it would be a bad place to do this init processing!).

jmarshall commented:
Thanks for reviewing, @kevinjacobs-progenity. I've finally got back to this, and I think made all the changes we discussed:

* Reworked the optional argument handling so as to be ready for settings coming from the HTSFile in future;

* Turns out this was failing on Python 2.7 because reusing `force_str()` wasn't really the right thing to do. Instead I've added another function that always decodes UTF-8, even on Python 2 — which is appropriate for these VCF fields.

kevinjacobs-progenity commented:
@jmarshall : Many thanks!  I'll review again over the weekend. 

On a related note, I've had several user requests to add automatic URL encoding and decoding to escape characters in INFO and GT string values.  It seems like it would be relatively easy to do this, building on this work.  Does this make sense to do here or should we wait until htslib adds this functionality?

kevinjacobs-progenity commented:
Arg.  I completely forgot I _had_ a working branch for HTSlib with percent escaping of string values.  I was changing jobs and I must be getting a bit senile.  I'll see how much of that original work is still relevant and decide how to proceed.

olingerc commented:
Sorry to bump this again. Do you have any news about this? I also haven't found any new release since 0.16.0.1. Is master safe to use?

kevinjacobs-progenity commented:
The main branch should be safe to use.  And I've lightly tested this branch.  However, life is currently complicated and I have not made substantive progress on implementing finer-grained control of encoding.  I'll set aside time in the next few days to give this PR a more thorough review so that it can be merged.

jmarshall commented:
I think this branch is itself feature-complete and is a good base for future work on finer-grained control of encoding. And it has demonstrated [test passes with all Python versions](https://github.com/jmarshall/pysam/actions/runs/820214292).

There's only so many balls I want to have in the air at once, and this is now holding up other improvements. @kevinjacobs-progenity: I'll aim to merge this at the end of the week if there are no objections (and if nobody else has already merged it by then).

kevinjacobs-progenity commented:
I've done some more light testing, reviewed the code, and am fine with merging.
